### DISCUSSION TITLE
Array range implementation.

### DISCUSSION THREAD

**Discussion by mertcandav** [dotnet/csharplang#4274] [Language Ideas]
I think it would be good to add a range implementation to C#. The return type is always a static array.
Suppose we define an array like this:
```csharp
int[] array = { 0, 1, 2, 3, 4, 5 };
```
The description with the range I think of in my mind:
```csharp
int[] array = 0:5; // Result: 0, 1, 2, 3, 4, 5
```
I think this could be an improvement that will make our work a lot easier in big definitions and different usage laws.
Likewise, if we want the range to increase by 2, for example, a derivative can be introduced as follows:
```csharp
int[] array = 0:5:2; // Result: 0, 2, 4
```
It can also be applied to char type:
```csharp
char[] letters = 'A':'Z';
```
I think it will be a good improvement in terms of readability and time savings. Please specify if you have a more logical syntax or idea. I think you understand the idea.

**Comment by huoyaoyuan** [comment]
> good improvement in terms of readability

It decreases readability, because weather the index is inclusive is unclear.

With a method, the code reader can get clear understanding from the documentation of the method.

**Reply by msedi** [reply]
> It decreases readability, because weather the index is inclusive is unclear.

I wouldn't agree on that. There are many things in .NET where it's not always quite clear if the index is exlusive or not, e.g.

- `Parallel.For`, you can only see it from the arguments (fromInclusive, toExclusive)
- `Partitioner`, I always have to read the docs if the Item2 is inclusive or not
- `Range`, End is exclusive, but could also be inclusive

I only want to express that it's now always defined, obviously per convention the methods described above follow the convention that the end is always exclusive but you can only be sure if it's described either by the parameter, the property name or the docs.

I would say that this is the same here.

**Comment by svick** [comment]
I also don't understand how does this improve the situation over a regular method with the same functionality. For example:

```c#
int[] array = Ranges.Array(0, 5); // Result: 0, 1, 2, 3, 4, 5
int[] array = Ranges.Array(0, 5, step: 2); // Result: 0, 2, 4; the "step:" part is optional
char[] letters = Ranges.Array('A', 'Z');
```

I think it's not something that's used that often, so it doesn't need to be super succinct.

**Reply by msedi** [reply]
> I think it's not something that's used that often, so it doesn't need to be super succinct.

I cannot completely agree on this statement. I think it depends on the area you are working on. If you do work a lot with image processing or data analytics you need this very often (e.g. for interpolation or down/upsampling). There was once a discussion around `Range `and `Index `when they were about to be introduced. But I couldn't find the discussion anymore.

There people also discussed the syntax. From what I can see @mertcandav describes some similar syntax as matlab uses it. I think python has some equivalent syntax.

When you are doing this many times your code gets cluttered with code that make the core of the "equation" unreadable. Something similar was there a whole ago until `using static` was introduced. Here we many code parts with Math.Sin/Cos, etc. methods where `using static` helped making the expression/equation more readable.

When `Range` was introduced one could also argue here, everything you can you with `Range `could have also been expression by `Range.Create()`. I'm not sure how widely the Range .. syntax has been accepted by users. I'm even not sure if I like it but I also think it makes code more readable.

**Comment by hez2010** [comment]
You can achieve similar code with extension methods:

https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AXEUCuA7AHwAEAmARgFgAoIgBgAIiyUBua6ogZkbKUZPoAlAIZ4A5jACiCDDDwBnAJYQF1AN7V6WxtyZ8AkpLw4AtjCjCM0ADyK8GAHz0A4jAxHT5y9AAUGABaK8kKiEvQW4jAAlJraGlTaifSKAGb0PhESAHQAyhjCUBhZ+vIAYlAQJkYAJvQEBOGhMFk1xWUVVXjVMQlJWvF9fQEVAO70eDBjAHIQGDk4AA4L0LLV0mAwCxjKeD5RbL19AL6xfadJqemZzXkFRQBqwgA2ODD01o2RLV1Zjy/R50SA0GiRS0HSADcCsl6ABeT7ZW6FX7PV4sGEfa7faoo/7oxQAagJPRBQMBpK0TDIjAA7MkDhStCdDklmSCYE95DByf0eUkwVBIdDFHCETd8si/miYU4sa0pTB8XA4CTGcDGdoqbT6XzEmzBvqmedzlweHw7BgANoAXXoABUIABBKAWACefkCwREkTFqt5LMSUMFTyCGFFEzGABlQ7Z7A49gzBgKYMIwP4hYKRXZfXz1aSQ/Iio7qtUfIp9nzDUk+UQ6QWig7nW6E+dmcyOLoBABhdQmkgkc55wPQgpu0U+WhZLKcKJZRsu4TuisBzUkACso8Xy711COQA==
